cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java

```diff
@@ -41,4 +43,37 @@ public Map<String, String> getProperties() {
         .map(HashMap::new)
         .orElseGet(HashMap::new);
   }
+
+  /**
+   * Allocates Table Space for the storage.
+   *
+   * <p>Default tableLocation looks like: {endpoint}/{rootPrefix}/{databaseId}/{tableId}-{tableUUID}
+   *
+   * @return the table location where the table data should be stored
+   */
+  @Override
+  public String allocateTableSpace(
+      String databaseId,
+      String tableId,
+      String tableUUID,
+      String tableCreator,
+      boolean skipProvisioning) {
+    Preconditions.checkArgument(databaseId != null, "Database ID cannot be null");
+    Preconditions.checkArgument(tableId != null, "Table ID cannot be null");
+    Preconditions.checkArgument(tableUUID != null, "Table UUID cannot be null");
+    Preconditions.checkState(
+        storageProperties.getTypes().containsKey(getType().getValue()),
+        "Storage properties doesn't contain type: " + getType().getValue());
+    return URI.create(
+            getClient().getEndpoint()
```

Comment:
How does URI.create or normalization handle situations where you have double slashes? For example, if the following endpoint and rootPrefix were provided:
Endpoint: hdfs://localhost:8080/
Rootprefix: /userFoo/

This would pass in the string: "hdfs://localhost:8080//userFoo//databaseId/tableId-tableUUID"

Are we relying on URI.create to do the validation of the user supplied endpoint and rootPrefix? If so, we might want to add some unit tests covering what we expect the output string to be.

In a perfect world, the following configurations would all produce the same location (i.e. the strings produced by this method are equal). I am not sure if these are all covered by normalization:

Config1:
Endpoint: hdfs://localhost:8080/
Rootprefix: /userFoo/

Config2:
Endpoint: hdfs://localhost:8080
Rootprefix: /userFoo

Config3:
Endpoint: hdfs://localhost:8080
Rootprefix: /userFoo/

Config4:
Endpoint: hdfs://localhost:8080/
Rootprefix: /userFoo

Also, is this considered an invalid configuration?
Config5:
Endpoint: hdfs://localhost:8080/
Rootprefix: userFoo

Reply:
Nice test cases! let me add them. I think current implementation should yield same result for all of them. 

 Normalize removes double slashes besides the scheme. 
Config 5 seem valid and likely what we should use for S3. RootPrefix should just be bucketName.

Reply:
Added tests and comments like discussed. This one is not the underlying assumption is endpoint+rootprefix should be valid.